lib/copy/copy.js

```diff
@@ -2,51 +2,213 @@
+        return onFile(src, srcStat, dest, options, cb)
+      } else if (srcStat.isSymbolicLink()) {
+        return onLink(src, dest, options, cb)
+      }
+    })
+  })
+}
+
+function checkDest (dest, cb) {
+  fs.readlink(dest, (err, resolvedDestPath) => {
+    if (err) {
+      if (err.code === 'ENOENT') return cb(null, DEST_NOENT)
+      // dest exists but is not a link
+      else if (err.code === 'EINVAL' || err.code === 'UNKNOWN') return cb(null, DEST_EXISTS)
+      else return cb(err)
+    } else { // dest exists and is a link
+      return cb(null, resolvedDestPath)
     }
+  })
+}
+
+function onFile (src, srcStat, dest, options, cb) {
+  checkDest(dest, (err, res) => {
+    if (err) return cb(err)
+    if (res === DEST_NOENT) {
+      return cpFile(src, srcStat, dest, options, cb)
+    } else if (res === DEST_EXISTS) {
+      if (options.overwrite) {
+        fs.unlink(dest, err => {
+          if (err) return cb(err)
+          return cpFile(src, srcStat, dest, options, cb)
+        })
+      } else if (options.errorOnExist) {
+        return cb(new Error(dest + ' already exists'))
+      } else return cb()
+    } else return cb()
+  })
+}
+
+function cpFile (src, srcStat, dest, options, cb) {
+  const rs = fs.createReadStream(src)
+  const ws = fs.createWriteStream(dest, { mode: srcStat.mode })
 
-    fs.exists(dir, dirExists => {
-      if (dirExists) return ncp(src, dest, options, callback)
-      mkdir.mkdirs(dir, err => {
-        if (err) return callback(err)
-        ncp(src, dest, options, callback)
+  rs.on('error', err => cb(err))
+  ws.on('error', err => cb(err))
+
+  ws.on('open', () => {
+    rs.pipe(ws)
+  }).once('close', () => {
+    fs.chmod(dest, srcStat.mode, err => {
+      if (err) return cb(err)
+      if (options.preserveTimestamps) {
+        return utimes(dest, srcStat.atime, srcStat.mtime, cb)
+      } else return cb()
+    })
+  })
+}
+
+function onDir (src, srcStat, dest, options, cb) {
+  checkDest(dest, (err, res) => {
+    if (err) return cb(err)
+    if (res === DEST_NOENT) {
+      // if dest is a subdir of src, prevent copying into itself
+      if (isSrcSubdir(src, dest)) return cb(new Error(`Cannot copy directory '${src}' into itself '${dest}'`))
+      fs.mkdir(dest, srcStat.mode, err => {
+        if (err) return cb(err)
+        fs.chmod(dest, srcStat.mode, err => {
+          if (err) return cb(err)
+          return cpDir(src, dest, options, cb)
+        })
       })
+    } else if (res === DEST_EXISTS) {
+      if (isSrcSubdir(src, dest)) return cb(new Error(`Cannot copy directory '${src}' into itself '${dest}'`))
+      return cpDir(src, dest, options, cb)
+    } else if (res && typeof res !== 'number') { // dest exists and is a link
+      if (isSrcSubdir(src, res)) return cb(new Error(`Cannot copy directory '${src}' into itself '${res}'`))
+      if (src === res) return cb()
+      return cpDir(src, dest, options, cb)
+    } else return cb()
```

Comment:
If you mean line 141 `} else return cb()`, I guess it is not needed and should be removed as we check for all possible conditions.